terraform/hcl2json_test.go

```diff
@@ -561,20 +561,171 @@ variable "test" {
 }`,
 			expected: VariableMap{
 				"var": cty.ObjectVal(VariableMap{
-					"dummy": cty.StringVal("dummy"),
+					"dummy": cty.StringVal("dummy_value"),
 				}),
 				"local": cty.ObjectVal(VariableMap{
-					"dummy": cty.StringVal("dummy"),
+					"dummy": cty.StringVal("dummy_local"),
 				}),
 			},
 		},
 	}
 
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
-			actual, err := ExtractVariables("test", tc.input)
+			actual, err := extractVariables("test", tc.input)
 			require.Nil(t, err)
 			assert.Equal(t, tc.expected, actual)
 		})
 	}
 }
+
+func TestParseModuleSuccess(t *testing.T) {
+	fileContent := `
+resource "aws_security_group" "allow_ssh" {
+	name        = "allow_ssh"
+	description = "Allow SSH inbound from anywhere"
+	cidr_blocks = var.dummy
+}`
+	jsonOutput := `{
+	"resource": {
+		"aws_security_group": {
+			"allow_ssh": {
+				"cidr_blocks": "dummy_value",
+				"description": "Allow SSH inbound from anywhere",
+				"name": "allow_ssh"
+			}
+		}
+	}
+}`
+	type test struct {
+		name       string
+		files      map[string]interface{}
+		extractErr *CustomError
+		parseErr   *CustomError
+		expected   map[string]interface{}
+	}
+	tests := []test{
+		{
+			name: "Multiple valid files with dummy variables and no error",
+			files: map[string]interface{}{
+				"test1.tf": fileContent,
+				"test2.tf": fileContent,
+			},
+			expected: map[string]interface{}{
+				"failedFiles": map[string]interface{}{},
+				"parsedFiles": map[string]interface{}{
+					"test1.tf": jsonOutput,
+					"test2.tf": jsonOutput,
+				},
+			},
+		},
+		{
+			name: "Multiple files and one file with a user error at extract time",
+			files: map[string]interface{}{
+				"fail.tf":  fileContent,
+				"test2.tf": fileContent,
+			},
+			extractErr: &CustomError{
+				message:   "User error",
+				errors:    []error{},
+				userError: true,
+			},
+			expected: map[string]interface{}{
+				"failedFiles": map[string]interface{}{
+					"fail.tf": "User error",
+				},
+				"parsedFiles": map[string]interface{}{
+					"test2.tf": jsonOutput,
+				},
+			},
+		},
+		{
+			name: "Multiple files and one file with an internal error at extract time",
+			files: map[string]interface{}{
+				"fail.tf":  fileContent,
+				"test2.tf": fileContent,
+			},
+			extractErr: &CustomError{
+				message:   "Internal error",
+				errors:    []error{},
+				userError: false,
+			},
+			expected: map[string]interface{}{
+				"failedFiles": map[string]interface{}{},
+				"parsedFiles": map[string]interface{}{
+					"fail.tf":  jsonOutput, // it's intentional for files that fail with internal errors at extraction time to still try to parse
```

Comment:
correct  me if I'm wrong:
user errors are errors that happened because of bad hcl, for example, but internal errors can be anything else that happened during parsing. is that right?
I'm curious  about this comment, what does it mean? if a file failed with an internal error when it was under the extractVariables method, we will still try to parse it? don't we skip it when we are checking `if failedFiles[fileName] == nil`?

Reply:
So at the moment `failedFiles` contains errors we want to show the user - internal errors don't get shown to users at all. I'm interested to see if you think this is a good idea or not btw

When we iterate through the files to extract variables, we check if there's an error and if there is one:
- and it's a user error, we include it in `failedFiles` (i.e. `failedFiles[fileName] != nil`) and don't push anything to the array of `VariableMap`s
- and it's an internal error, we don't include it in `failedFiles` (see first paragraph in this comment) and don't push anything to the array of `VariableMap`s
Then, we iterate again through the files to parse them (this step), but only if there weren't any user errors in the previous step to extract variables (i.e. `failedFiles[fileName] == nil`). So we still try to parse them if there was an internal error in case there's a tiny chance the error was a flake.

Does this make sense? I can add a comment to the `ParseModule` function to explain this a bit, but will wait to see what you think

Reply:
yes, thanks, I think it hadn't clicked to that we can have "flakes". have you seen examples of this before I assume?

for a comment, even if you update this line 
"// it's intentional for files that fail with internal errors at extraction time to still try to parse" with something extra is still good e.g.
 "// it's intentional for files that fail with internal errors at extraction time to still try to parse as the internal error can be a flake"

Reply:
I was thinking about some of the feedback we'd been getting during usability testing - see https://snyk.slack.com/archives/C02JMMTLUF9/p1643647684197409. At the moment we print out a lot of errors that the user can't really action, so this separation was meant to help with that.